From ce77469bb903583bbde78f20639063bac7c107fc Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Wed, 20 Dec 2023 10:48:18 +0800
Subject: [PATCH 16/99] gl-renderer: Fix width/stride error for FBO

Tested on RK3588 EVB with:
weston -B vnc-backend.so --renderer=gl

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 libweston/renderer-gl/gl-renderer.c | 16 +++++++++++++---
 1 file changed, 13 insertions(+), 3 deletions(-)

diff --git a/libweston/renderer-gl/gl-renderer.c b/libweston/renderer-gl/gl-renderer.c
index 2d98cc7..430df40 100644
--- a/libweston/renderer-gl/gl-renderer.c
+++ b/libweston/renderer-gl/gl-renderer.c
@@ -796,6 +796,7 @@ gl_renderer_do_read_pixels(struct gl_renderer *gr,
 {
 	void *read_target;
 	pixman_image_t *tmp = NULL;
+	void *tmp_data = NULL;
 
 	assert(fmt->gl_type != 0);
 	assert(fmt->gl_format != 0);
@@ -811,11 +812,17 @@ gl_renderer_do_read_pixels(struct gl_renderer *gr,
 		 * glReadPixels() returns bottom row first. We need to
 		 * read into a temporary buffer and y-flip it.
 		 */
+		tmp_data = malloc(stride * rect->height);
+		if (!tmp_data)
+			return false;
+
 		tmp = pixman_image_create_bits(fmt->pixman_format,
 					       rect->width, rect->height,
-					       NULL, 0);
-		if (!tmp)
+					       tmp_data, stride);
+		if (!tmp) {
+			free(tmp_data);
 			return false;
+		}
 
 		read_target = pixman_image_get_data(tmp);
 	}
@@ -850,6 +857,7 @@ gl_renderer_do_read_pixels(struct gl_renderer *gr,
 
 		pixman_image_unref(image);
 		pixman_image_unref(tmp);
+		free(tmp_data);
 	}
 
 	return true;
@@ -2042,7 +2050,9 @@ gl_renderer_repaint_output(struct weston_output *output,
 			pixels += extents->x1 - (int)output->pos.c.x;
 		}
 
-		gl_renderer_do_read_pixels(gr, compositor->read_format, pixels, stride, &rect);
+		/* compositor->read_format is a 32b format */
+		gl_renderer_do_read_pixels(gr, compositor->read_format,
+					   pixels, stride * 4, &rect);
 
 		if (gr->gl_version >= gr_gl_version(3, 0))
 			glPixelStorei(GL_PACK_ROW_LENGTH, 0);
-- 
2.20.1

